docs: design for atlas pull and translations management | FC-0055 (#367)
openedx/openedx-app-ios: docs/0002-atlas-translations-management.rst

Atlas Translations Management Design
####################################

Date: 25 March 2024

Status
******
Pending

Context
*******

Open edX microservices and micro-frontends is using `OEP-58`_ to pull the latest
translations from the `openedx-translations repository`_.

> The gist of `OEP-58`_ is two processes:
> 
> - Extract the English translation source entries and commit them into the `openedx-translations repository`_
> - Provide a command to pull translations in both development and web server image builds

Author: ```suggestion
The main changes that `OEP-58`_ introduced to the Open edX project are the following:

- Extract the English translation source entries and commit them into the `openedx-translations repository`_
- The `GitHub Transifex app`_ will automatically commit the translations to the `openedx-translations repository`_
- Provide a command that utilizes the `atlas`_ tool to pull translations in both development and web server image builds
```

> Microservices, micro-frontends, XBlocks and plugins uses one or two files which makes it easier for Translators

Reviewer: ```suggestion
Microservices, micro-frontends, XBlocks and plugins use files which make it easier for Translators
```

Author: ```suggestion
Microservices, micro-frontends, XBlocks and plugins uses as few translation files as possible to makes it easier for Translators
```

Reviewer: Ah! That wasn't clear to me from the previous wording. "As few as possible" sounds good!

Grammer nit: "use" not "uses" here

to locate and translate relevant resources.

> The `edx-platform`_ used to have 7 different resources, but this has been deprecated in favor of ``edx-platform.po``

Author: ```suggestion
`edx-platform`_ translations strings used to be split between 7 files, but this has been deprecated in favor of ``edx-platform.po``
```

and ``edx-platform-js.po`` files. The details in the `edx-platform resources combination decision`_.

> The ``openedx-app-ios`` uses a clean architecture. Therefore, each app has its own i18n strings files. This is
> useful for Developer Experience, but we believe it's not good for Translator Experience.

Reviewer: ```suggestion
In ``openedx-app-ios``, each app has its own i18n strings file. This is useful for
Developer Experience, but we believe it's not good for Translator Experience.
```

It'd probably be good to explain *why* this is a bad translator experience.

Decision
********

> This decision will require few changes to the translation workflow in comparison with the old ``edx-app-ios``:

Reviewer: ```suggestion
This decision requires a few changes to the mobile app translation workflow:
```

> - Translation files are no longer comitted directly in the mobie app ``openedx-app-ios/android`` repositories

Reviewer: ```suggestion
- Translation files are no longer comitted directly to the mobie app repositories (``openedx-app-ios``/``openedx-app-android``)
```

It might be nice to turn those repo names into links.

Author: ```suggestion
- Translation files are no longer comitted directly to the mobie app repositories (`openedx-app-ios`_ / `openedx-app-android`_)
```

> - Keep the i18n string files split into each clean architecture app without hurting Developer Experience.

Reviewer: I feel like there's probably a better term than "clean architecture app," but I'm not sure exactly what that is.

Author: Same here, but I don't know what should I use. The word "Project" can be a better alternative.

Reviewer: @volodymyr-chekyrta suggested "module" and I like that a lot!

- Identical string IDs across the application will be treated as the same string/translation.
  For example, ``BUTTON_CLOSE="Close"`` in dashboard is the same ``BUTTON_CLOSE="Close"`` in profile app.

>   Otherwise, a validation script will throw an error and prevent merging the pull request.

Reviewer: Would it be possible to namespace these? So, for example, when compiling the strings into a resource file we could have `dashboard.BUTTON_CLOSE` and `profile.BUTTON_CLOSE`?

I'd think ideally this could happen as part of the source string compilation process, so each section could still just use `BUTTON_CLOSE`, but if that isn't feasible then directly adding the namespace to the strings in each section could work too.

Author: It's possible, but we'd rather keep the strings intact to avoid introducing unintended bugs.

@volodymyr-chekyrta what do you think?

Author: The same idea has already been suggested by Volodymyr. I'll mark this as resolved.

https://github.com/openedx/openedx-app-ios/pull/367#discussion_r1540988521

Thanks again!

- String files are combined into a single file before preparing it to Transifex

>   (through the `openedx-translations repository`_)

Reviewer: Instead of "preparing it to" maybe "uploading to" or "syncing to"

Author: ```suggestion
- String files are combined into a single file before uploading it to Transifex
  (through the `openedx-translations repository`_)
```

> - Translators can translate the entire app as a single resource

Reviewer: ```suggestion
- Translators can translate the entire app within a single transifex resource
```

- Before releasing to App Store, developers need to run a new ``make pull_translations`` command

> - After translations is pulled the translation file is split again into each clean architecture application
>   as if it has been individual resource (a new script)

Reviewer: ```suggestion
- After translations are pulled the translation files are split into each clean architecture application
  as if it has been individual resource (a new script)
```

> String files combination and splitting process
> ==============================================

Reviewer: ```suggestion
String file combination and splitting process
=============================================
```

while Engineers need to have clean architecture with strings split into each app to aid maintainability.

Combining the files during ``combine_translations``
---------------------------------------------------

However, combining the string files to aid Translator Experience in the following manner:

Suppose we have two apps with one string source file in each::

  Course/Course/en.lproj/Localizable.strings
  Dashboard/Dashboard/en.lproj/Localizable.strings

This script will collect the strings, remove uneeded comments and combine them in a temporary file::

  I18N/I18N/en.lproj/Localizable.strings

This file will be committed into the `openedx-translations repository`_ as described in `OEP-58`_.

This process happens entirely on the CI server (GitHub in this case) after each pull request merge without developer
intervention.

GitHub Actions CI Script to avoid duplicate string IDs
------------------------------------------------------

A new script will be created to ensure we don't have duplicate IDs with different content.

For example:

If the Dashboard has ``BUTTON_CLOSE="Close"`` while Course app has ``BUTTON_CLOSE="CLOSE TAB"``, this

> script should fail and require either making the string to match or the ID to be different.

Author: @volodymyr-chekyrta @brian-smith-tcril any objections against this validation?

For example I see `TITLE` being used multiple times. Any reason it can't be `DASHBOARD.TITLE` and `WHATS_NEW.TITLE`?

https://github.com/openedx/openedx-app-ios/blob/05d3073a822ed12783140aa2744dca939185fe08/Discussion/Discussion/en.lproj/Localizable.strings#L9

https://github.com/openedx/openedx-app-ios/blob/05d3073a822ed12783140aa2744dca939185fe08/WhatsNew/WhatsNew/en.lproj/Localizable.strings#L9

If that's okay, then I believe this part is good to go.

Reviewer 2: Good question.
Duplication of keys in modules is allowed because each module is a separate independent project/service, so it's not a problem for development.
And introducing a new rule for naming could lead to more complexity in the project and more time for onboarding for newcomers, I'd like to avoid it.

Is there any change to store each module's resources separately in their own files without merging them?
It seems that the edX app translations to Transifex are stored in different files with no problem. 
I wonder if we can keep the resources separate.

<img width="1029" alt="image" src="https://github.com/openedx/openedx-app-ios/assets/127732735/9251eba5-22c7-4f70-8416-1deec7347a98">

Author: > Good question.
> Duplication of keys in modules is allowed because each module is a separate independent project/service, so it's not a problem for development.

Yes, this is very well understood. We will try to not affect clean architecture apps when it comes to developer experience. 

> And introducing a new rule for naming could lead to more complexity in the project and more time for onboarding for newcomers, I'd like to avoid it.

I'll explain what we intend:

 - It's perfectly fine to have duplicate strings that are identical .e.g having `"BUTTON_CLOSE" = "Close"` in multiple apps is fine. 
 - Engineers can focus on their app unaware of the other apps.
 - The only exception is that we'll need to avoid having the same ID used for different strings:

For example the iOS app has only the following strings need fixing:

```
Duplicate IDs with different messages:
ID: "ALERT.ACCEPT"
  Message 1: "Accept";
  Message 2: "ACCEPT";
ID: "TITLE"
  Message 1: "Dashboard";
  Message 2: "Profile";
  Message 3: "Discussions";
  Message 4: "Discover";
  Message 5: "What's New";
ID: "SEARCH.EMPTY_DESCRIPTION"
  Message 1: "Start typing to find the topics";
  Message 2: "Start typing to find the course";
```

While other IDs are completely fine:

```
Duplicate IDs with different messages:
ID: "COURSEWARE.COURSE_CONTENT"
  Message 1: "Course content";
  Message 2: "Course content";
ID: "COURSEWARE.COURSE_UNITS"
  Message 1: "Course units";
  Message 2: "Course units";
ID: "COURSEWARE.NEXT"
  Message 1: "Next";
  Message 2: "Next";
ID: "COURSEWARE.PREVIOUS"
  Message 1: "Prev";
  Message 2: "Prev";
ID: "COURSEWARE.FINISH"
  Message 1: "Finish";
  Message 2: "Finish";
ID: "COURSEWARE.GOOD_WORK"
  Message 1: "Good Work!";
  Message 2: "Good Work!";
ID: "COURSEWARE.BACK_TO_OUTLINE"
  Message 1: "Back to outline";
  Message 2: "Back to outline";
ID: "COURSEWARE.SECTION"
  Message 1: "Section “";
  Message 2: "Section “";
ID: "COURSEWARE.IS_FINISHED"
  Message 1: "“ is finished.";
  Message 2: "“ is finished.";
ID: "COURSEWARE.CONTINUE"
  Message 1: "Continue";
  Message 2: "Continue";
ID: "COURSEWARE.RESUME_WITH"
  Message 1: "Resume with:";
  Message 2: "Resume with:";
ID: "ERROR.RELOAD"
  Message 1: "Reload";
  Message 2: "Reload";
ID: "SEARCH"
  Message 1: "Search";
  Message 2: "Search";
ID: "SEARCH.TITLE"
  Message 1: "Search results";
  Message 2: "Search results";
```

I hope that makes it less of an issue to introduce this lint rule with clear GitHub comments via GitHub Actions it should be an acceptable minimal cost imho.

We can pursue a more complex alternative, but we'd like to start with something simple.

> Is there any change to store each module's resources separately in their own files without merging them?
It seems that the edX app translations to Transifex are stored in different files with no problem.
I wonder if we can keep the resources separate.

Yes, we'd like to change that because we think it will improve Translators Experience.

The new resources will have to be in the [openedx-translations Transifex project](https://app.transifex.com/open-edx/openedx-translations/content/) which currently has 57 resources and the number is growing. 

It's hard to justify adding 10 resources for each app, and require the mobile engineers to follow the [Enable Translations on a New Repository](https://docs.openedx.org/en/latest/developers/how-tos/enable-translations-new-repo.html#id24) every time a new app is created. It's quite of a process and we'd rather save the mobile engineers the effort with some automation.

This proposal ensures that mobile engineers continue with their workflow mostly uninterrupted. 

That was long, I hope it's clearer now.

Author: @brian-smith-tcril I'd like to have your input here. The decision here involves combining the resources into a single Transifex resource [similar to the edx-platform](https://github.com/openedx/edx-platform/blob/master/docs/decisions/0018-standarize-django-po-files.rst) to avoid adding too many resources to the transifex project.

It's worth noting that the scope of FC-55 is mostly focused on combining the resources to improve Translators Experience.

Please let me know what do you think a good trade-off is here.

Reviewer 2: @OmarIthawi, I fully understand that this is the only issue among the 9 strings for now. 
It doesn't seem like a big deal to rename them, we can do it, but maybe we can tweak it by adding a `temp_$module_name` prefix to those IDs to handle that via script?
I'm worried about adding a new naming rule because it could make the project more complex. I'd rather keep things simple and accessible for the open-source community.

Author: Thanks @volodymyr-chekyrta @brian-smith-tcril I think that's a pretty reasonable way to go.

To confirm, you're suggesting that we only prefix the duplicate strings, and keep the rest unchanged before pushing to Transifex. Right?

I'm asking this because I'm wondering if it makes sense to prefix all strings, and remove the prefix after splitting. 

Both options will remove the need for the lint rule.

Reviewer 2: Hmm 🤔 , maybe adding prefixes to all strings makes sense.
Form something like `$module_name + $string_id`.
@OmarIthawi @brian-smith-tcril what do you think? 
Will it simplify script development?

Author: Works for me and as far as I know, it'll be a nice translation experience if someone wants to know the code location of the resource.

I'll update the proposal.

Atlas Translations Management Design
####################################

Date: 25 March 2024

Status
******
Pending

Context
*******

Open edX microservices and micro-frontends use the processes outlined in `OEP-58`_ to pull the latest
translations from the `openedx-translations repository`_.

The main changes that `OEP-58`_ introduced to the Open edX project are the following:

- Extract the English translation source entries and commit them into the `openedx-translations repository`_
- The `GitHub Transifex app`_ will automatically commit the translations to the `openedx-translations repository`_
- Provide a command that utilizes the `atlas`_ tool to pull translations in both development and web server image builds

Microservices, micro-frontends, XBlocks and plugins use as few translation files as possible to makes it easier for Translators
to locate and translate relevant resources.

`edx-platform`_ translations strings used to be split between 7 files, but this has been deprecated in favor of
``edx-platform.po`` and ``edx-platform-js.po`` files.
More details are available in the `edx-platform resources combination decision document`_.

> The `openedx-app-ios`_ uses a clean architecture. Therefore, each app has its own i18n strings files. This is

Reviewer 2: Can we use term `module` instead of `app` or something similar?

Reviewer 2: Thank you!

  Otherwise, a validation script will throw an error and prevent merging the pull request.
- String files are combined into a single file before uploading it to Transifex
  (through the `openedx-translations repository`_)
- Translators can translate the entire app within a single transifex resource
- Before releasing to App Store, developers need to run a new ``make pull_translations`` command
- After translations are pulled the translation files are split into each clean architecture application
  as if it has been individual resource (a new script)

String file combination and splitting process
=============================================
This is a new process that's being introduced to have the best combination
of Developer Experience and Translator Experience.

Translators need to have less resources as possible with strings de-duplicated,
while Engineers need to have clean architecture with strings split into each app to aid maintainability.

Combining the files during ``combine_translations``
---------------------------------------------------

However, combining the string files to aid Translator Experience in the following manner:

> Suppose we have two apps with one string source file in each::

Reviewer 2: And here, can we use some other term instead of `apps`? It's confusing a bit.

Author: done

Atlas Translations Management Design
####################################

Date: 25 March 2024

Status
******
Accepted

Context
*******

Open edX microservices and micro-frontends use the processes outlined in `OEP-58`_ to pull the latest
translations from the `openedx-translations repository`_.

The main changes that `OEP-58`_ introduced to the Open edX project are the following:

- Extract the English translation source entries and commit them into the `openedx-translations repository`_
- The `GitHub Transifex app`_ will automatically commit the translations to the `openedx-translations repository`_
- Provide a command that utilizes the `atlas`_ tool to pull translations in both development and web server image builds

> Microservices, micro-frontends, XBlocks and plugins use as few translation files as possible to makes it easier for Translators

Reviewer: ```suggestion
Microservices, micro-frontends, XBlocks and plugins use as few translation files as possible to make it easier for Translators
```

to locate and translate relevant resources.

> `edx-platform`_ translations strings used to be split between 7 files, but this has been deprecated in favor of

Reviewer: ```suggestion
`edx-platform`_ translation strings used to be split between 7 files, but this has been deprecated in favor of
```

``edx-platform.po`` and ``edx-platform-js.po`` files.
More details are available in the `edx-platform resources combination decision document`_.

> The `openedx-app-ios`_ uses a clean architecture. Therefore, each clean architecture module has its own i18n
> localizable strings file. This is useful for Developer Experience.
> However, it's not good for Translator Experience because translators
> often are unaware of the application engineering architecture and would like to be able to quickly search for strings
> regardless of their location in the code.

Reviewer: ```suggestion
`openedx-app-ios`_ uses a modular architecture. Each module has its own i18n
localizable strings file. This is useful for Developer Experience, but can be harmful
to the Translator Experience. Translators often are unaware of the engineering
architecture and would like to be able to quickly search for strings regardless of
their location in the code.
```

Decision
********

This decision requires a few changes to the mobile app translation workflow:

- Translation files are no longer comitted directly to the mobie app repositories (`openedx-app-ios`_ / `openedx-app-android`_)
- Keep the i18n string files split into each module without hurting Developer Experience.
- String files are combined into a single file before uploading it to Transifex
  (through the `openedx-translations repository`_)
- Translators can translate the entire app within a single transifex resource
- Before releasing to App Store, developers need to run a new ``make pull_translations`` command
- After translations are pulled the translation files are split into each module

>   as if it has been individual resource (a new script)

Reviewer: This section is a little confusing. Some bullet points are describing changes to the workflow, some are describing ways in which the workflow will not change, and some are just describing the proposed new workflow without reference to anything changing/staying the same.

It might be helpful to split this up into a couple subsections:
```
New Workflow
* Bullets
* Describing
* New
* Workflow
```
```
Notable changes
* Bullets describing notable ways in which the new workflow is different from the previous workflow
```

Or maybe it could all be done in one section focusing on the new workflow, with sub-bullets under any bullets that include notable changes.

String file combination and splitting process
=============================================
This is a new process that's being introduced to have the best combination
of Developer Experience and Translator Experience.

> Translators need to have less resources as possible,
> while Engineers need to have clean architecture with strings split into each app to aid maintainability.

Reviewer: ```suggestion
The best experience for Translators requires combining source strings into as few transifex resources as possible.
The best experience for Engineers requires splitting translation source files to fit within the modular architecture.
```

Combining the files during ``combine_translations``
---------------------------------------------------

Combining the string files to aid Translator Experience in the following manner:

Suppose we have two modules with one Localizable.strings file in each::

  Course/Course/en.lproj/Localizable.strings
    "COURSE_CONTAINER.VIDEOS" =  "Videos";
    "COURSE_CONTAINER.DATES" =  "Dates";
    "COURSE_CONTAINER.DISCUSSIONS" =  "Discussions";
    "COURSE_CONTAINER.HANDOUTS" =  "Handouts";
    "COURSE_CONTAINER.HANDOUTS_IN_DEVELOPING" =  "Handouts In developing";

    "HANDOUTS_CELL_HANDOUTS.TITLE" = "Handouts";
    "HANDOUTS_CELL_ANNOUNCEMENTS.TITLE" = "Announcements";

  Dashboard/Dashboard/en.lproj/Localizable.strings:
    "TITLE" = "Dashboard";
    "HEADER.COURSES" = "Courses";
    "HEADER.WELCOME_BACK" = "Welcome back. Let's keep learning.";

    "EMPTY.TITLE" = "It's empty";
    "EMPTY.SUBTITLE" = "You are not enrolled in any courses yet.";

This combine translations script will collect the strings, remove unneeded comments and combine them in a temporary
file while prefixing each entry with its module name. This will make splitting clear and

> avoid collision in string IDs::

Reviewer: I know you meant code comments when saying "remove unneeded comments", but it reminded me of https://github.com/openedx/frontend-build/pull/460

Do we have a plan in place to support message descriptions for translators in transifex?

Author: I'm unware of how Translators comments work in iOS, but definilty it should be possible to support those comments.

There's no text processing limitation that would prevent us from pushing the comments as long as they're present in the `en.lproj/Localizable.strings`

More details are available in the `edx-platform resources combination decision document`_.

`openedx-app-ios`_ uses a modular architecture. Each module has its own i18n
localizable strings file. This is useful for Developer Experience, but can be harmful
to the Translator Experience. Translators often are unaware of the engineering
architecture and would like to be able to quickly search for strings regardless of
their location in the code.

Decision
********

Combine English ``Localizable.strings`` files from mobile app modules before pushing them to the `openedx-translations repository`_ and
split back the translated files into the modules after pulling from the `openedx-translations repository`_ via
`atlas`_.

Notable changes
===============

- Translation files are no longer committed directly to the mobie app repositories (`openedx-app-ios`_ / `openedx-app-android`_)
- String files are combined into a single file before uploading it to Transifex (through the `openedx-translations repository`_)
- Translators can translate the entire mobile app strings in a single Transifex resource
- Before releasing to App Store, developers need to run a new ``make pull_translations`` command

> - After translations are pulled the translation files are split into each module
>   as if it has been individual resource (a new script)

Reviewer: ```suggestion
- After translations are pulled the translation files are split by module
```

I'm not sure if we need to mention the individual resource stuff here

to locate and translate relevant resources.

`edx-platform`_ translation strings used to be split between 7 files, but this has been deprecated in favor of
``edx-platform.po`` and ``edx-platform-js.po`` files.
More details are available in the `edx-platform resources combination decision document`_.

`openedx-app-ios`_ uses a modular architecture. Each module has its own i18n
localizable strings file. This is useful for Developer Experience, but can be harmful
to the Translator Experience. Translators often are unaware of the engineering
architecture and would like to be able to quickly search for strings regardless of
their location in the code.

Decision
********

Combine English ``Localizable.strings`` files from mobile app modules before pushing them to the `openedx-translations repository`_.
split back the translated files into the modules after pulling from the `openedx-translations repository`_ via
`atlas`_.

The `OEP-58`_ workflow will proceed as follows:

- The `openedx-translations repository`_ runs a daily cronjob to collect strings from the `openedx-app-ios`_
  ``main`` branch.

>   - The `extract-translation-source-files.yml`_ workflow will clone the `openedx-app-ios`_.

Reviewer: ```suggestion
  - The `extract-translation-source-files.yml`_ workflow will clone the `openedx-app-ios`_ repo.
```

>   - The `extract-translation-source-files.yml`_ workflow will run the `openedx-app-ios`_'s ``make combine_translations``.

Reviewer: ```suggestion
  - The `extract-translation-source-files.yml`_ workflow will run `openedx-app-ios`_'s ``make combine_translations``.
```

>   - The ``make combine_translations`` combines English ``Localizable.strings`` files from the mobile app modules into

Reviewer: ```suggestion
  - ``make combine_translations`` combines English ``Localizable.strings`` files from the mobile app modules into
```

    a single ``I18N/I18N/en.lproj/Localizable.strings`` file.

>   - The ``I18N/I18N/en.lproj/Localizable.strings`` is committed in the `openedx-translations repository`_

Reviewer: ```suggestion
  - The ``I18N/I18N/en.lproj/Localizable.strings`` file is committed in the `openedx-translations repository`_
```

  - The `GitHub Transifex app`_ will fetch the updated English source ``.strings`` files from `openedx-translations repository`_
  - Translators can translate the strings in the `Transifex openedx-translations project`_
  - The `GitHub Transifex app`_ syncs the translated strings into the `openedx-translations repository`_
- Developers run ``make pull_translations`` to pull the translations from the `openedx-translations repository`_

>   - The ``make pull_translations`` run `atlas pull`_ command pulls the translations

Reviewer: ```suggestion
  - ``make pull_translations`` runs the `atlas pull`_ command to pull the translations
```

>   - The translation will be pulled in a single file for each language e.g. ``I18N/I18N/ar.lproj/Localizable.strings``

Reviewer: ```suggestion
  - The translations will be pulled as a single file for each language e.g. ``I18N/I18N/ar.lproj/Localizable.strings``
```

  * The `extract-translation-source-files.yml`_ workflow will clone the `openedx-app-ios`_.

  * The `extract-translation-source-files.yml`_ workflow will run the `openedx-app-ios`_'s ``make combine_translations``.

  * The ``make combine_translations`` combines English ``Localizable.strings`` files from the mobile app modules into a single ``I18N/I18N/en.lproj/Localizable.strings`` file.

  * The ``I18N/I18N/en.lproj/Localizable.strings`` is committed in the `openedx-translations repository`_.

  * The `GitHub Transifex app`_ will fetch the updated English source ``.strings`` files from `openedx-translations repository`_

  * Translators can translate the strings in the `Transifex openedx-translations project`_

  * The `GitHub Transifex app`_ syncs the translated strings into the `openedx-translations repository`_

* Developers run ``make pull_translations`` to pull the translations from the `openedx-translations repository`_

  * The ``make pull_translations`` run `atlas pull`_ command pulls the translations

  * The translation will be pulled in a single file for each language e.g. ``I18N/I18N/ar.lproj/Localizable.strings``

>   * Then ``make pull_translations`` will run ``python scripts/split_translation_files.py`` script to split the translations into the modules e.g. ``Course/Course/ar.lproj/Localizable.strings``

Reviewer: ```suggestion
  * Then ``make pull_translations`` will run ``python scripts/split_translation_files.py`` to split
    the translations by module and place them in their corresponding directories
    e.g. ``Course/Course/ar.lproj/Localizable.strings``
```

`atlas`_.

The `OEP-58`_ workflow for mobile apps will proceed as follows:

* The `openedx-translations repository`_ runs a daily cronjob to collect strings from the `openedx-app-ios`_ ``main`` branch.

  * The `extract-translation-source-files.yml`_ workflow will clone the `openedx-app-ios`_ repo.

  * The `extract-translation-source-files.yml`_ workflow will run `openedx-app-ios`_'s ``make combine_translations``.

  * ``make combine_translations`` combines English ``Localizable.strings`` files from the mobile app modules into a single ``I18N/I18N/en.lproj/Localizable.strings`` file.

  * The ``I18N/I18N/en.lproj/Localizable.strings`` file is committed in the `openedx-translations repository`_.

  * The `GitHub Transifex app`_ will fetch the updated English source ``.strings`` files from `openedx-translations repository`_

  * Translators can translate the strings in the `Transifex openedx-translations project`_

  * The `GitHub Transifex app`_ syncs the translated strings into the `openedx-translations repository`_

* Developers run ``make pull_translations`` to pull the translations from the `openedx-translations repository`_

>   * ``make pull_translations`` runs `atlas pull`_ command to pull the translations

Reviewer: ```suggestion
  * ``make pull_translations`` runs the `atlas pull`_ command to pull the translations
```

or

```suggestion
  * ``make pull_translations`` runs `atlas pull`_ to pull the translations
```
